inductor: make fallback for cpu scatter_add

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108330
* __->__ #108220


For inductor cpu backend, the scatter_add will use ```atomic_add```, which get a worse performance, currently, we make fallback for it to avoid performance regression compared with eager mode(single socket of SKX):
```
basic_gnn_gin 1.16x(after) Vs 0.509x(before)

basic_gnn_sage  1.064x(after) Vs 0.496x (before)

basic_gnn_gcn 1.373x(aftre) Vs 0.720x(before)
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

XiaobingSuper committed:
make fallback for cpu scatter_add

[ghstack-poisoned]

XiaobingSuper committed:
Update on "make fallback for cpu scatter_add"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

jgong5 commented on torch/_inductor/lowering.py:
```diff
@@ -2764,6 +2764,7 @@ def scatter_fallback(
     reduce_ty = "add" if fn == "aten.scatter_" else "sum"
     if (
         reduce not in {None, reduce_ty}
+        or (reduce == "sum" and self.get_device() == torch.device("cpu"))
```
Can we add a configuration to opt in the codegen instead of fallback? With this, the current codegen for reduction with atomic-add can still be exercised.

XiaobingSuper replied:
Yes, added.

XiaobingSuper committed:
Update on "inductor: make fallback for cpu scatter_add"



For inductor cpu backend, the scatter_add will use ```atomic_add```, which get a worse performance, currently, we make fallback for it to avoid performance regression compared with eager mode(single socket of SKX):
```
basic_gnn_gin 1.16x(after) Vs 0.509x(before)

basic_gnn_sage  1.064x(after) Vs 0.496x (before)

basic_gnn_gcn 1.373x(aftre) Vs 0.720x(before)
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge
